src/main/java/com/amazon/opendistroforelasticsearch/sql/executor/format/SelectResultSet.java

```diff
@@ -366,9 +379,55 @@ private String getFieldName(Field field) {
             }
         }
 
+        if (isSelectAllOnly(query)) {
+            populateAllNestedFields(columns, fields);
+        }
+
         return columns;
     }
 
+    private boolean isFieldSpecifiedExplicitly(Map<String, Field> fieldMap, String fieldName) {
+        return fieldMap.containsKey(fieldName);
+    }
+
+    /**
+     *  Verify if field is property field (object or nested ) and matched by wildcard pattern given in SELECT
+     *  A special case is text field with nested keyword. Ignore it because it's a "hidden" field.
+     */
+    private boolean isFieldPropertyAndMatchWildcard(Map<String, Field> fieldMap, String fieldName) {
+        int lastDot = fieldName.lastIndexOf(".");
+        if (lastDot != -1) {
+            String path = fieldName.substring(0, lastDot);
+            return !fieldName.endsWith(".keyword") && fieldMap.containsKey(path + ".*");
+        }
+        return true;
+    }
+
+    /**
+     * SELECT * only without other columns or wildcard pattern specified.
+     */
+    private boolean isSelectAllOnly(Query query) {
+        return isSelectAll() && fetchFields(query).isEmpty();
+    }
+
+    /**
+     * Special case which trades off consistency of SELECT * meaning for more intuition from customer perspective.
+     * In other cases, * means all regular fields on the level.
+     * The only exception here is * picks all non-regular (nested) fields as JSON without flatten.
+     */
+    private void populateAllNestedFields(List<Schema.Column> columns, List<String> fields) {
+        Set<String> nestedFieldNames = fields.stream().
+                                       filter(f -> f.contains(".") && !f.endsWith("keyword")).
```

Comment:
could you introduce named constants and add more comments? For example - the condition here is endsWith("keyword") and the condition in line 401 is endsWith(".keyword"). That is subtle difference that could be explained by the comment or constant name.

Reply:
> could you introduce named constants and add more comments? For example - the condition here is endsWith("keyword") and the condition in line 401 is endsWith(".keyword"). That is subtle difference that could be explained by the comment or constant name.

Sorry, my mistake. It's suppose to be ".keyword" here too. Another example to support my proposal that build more solid domain layer rather than do this kind of tricky check. :) Will fix it.